src/conduits/java/com/enderio/conduits/common/types/EnergyConduitTicker.java

```diff
@@ -24,26 +27,37 @@ public EnergyConduitTicker() {
     public void tickCapabilityGraph(IConduitType<?> type, List<CapabilityConnection> inserts, List<CapabilityConnection> extracts, ServerLevel level,
         Graph<Mergeable.Dummy> graph, TriFunction<ServerLevel, BlockPos, ColorControl, Boolean> isRedstoneActive) {
 
-        int availableForExtraction = 0;
-        for (IEnergyStorage extract : extracts.stream().map(e -> e.cap).toList()) {
-            availableForExtraction += extract.extractEnergy(extract.getEnergyStored(), true);
-        }
+        toNextExtract:
+        for (CapabilityConnection extract : extracts) {
+            IEnergyStorage extractHandler = extract.cap;
+            int availableForExtraction = extractHandler.extractEnergy(Integer.MAX_VALUE, true);
+            if (availableForExtraction <= 0)
+                continue;
+            EnergyExtendedData.EnergySidedData sidedExtractData = extract.data.castTo(EnergyExtendedData.class).compute(extract.direction);
 
-        int inserted = 0;
-        for (IEnergyStorage insert : inserts.stream().map(e -> e.cap).toList()) {
-            inserted += insert.receiveEnergy(availableForExtraction - inserted, false);
-            if (inserted == availableForExtraction)
-                break;
-        }
+            if (inserts.size() <= sidedExtractData.rotatingIndex) {
+                sidedExtractData.rotatingIndex = 0;
+            }
 
-        for (IEnergyStorage extract : extracts.stream().map(e -> e.cap).toList()) {
-            inserted -= extract.extractEnergy(inserted, false);
-            if (inserted <= 0)
-                break;
-        }
+            for (int j = 0; j < sidedExtractData.rotatingIndex; j++) {
+                //empty lists are verified in ICapabilityAwareConduitTicker
+                //this moves the first element to the back to give a new cap the next time this is called
+                inserts.add(inserts.remove(0));
+            }
```

Comment:
instead of doing this add/remove thing, we should probably just do the second for loop as
for (int j = rotatingIndex; j < rotatingIndex + inserts.size; j++) and then modulo j by inserts.size, so we have less list shrinkages